Photo submissions to S3

redshiftzero opened the pull request:
This PR closes issue #5, to implement in-app photo uploads to S3. 

Submission route `/submit` is now available to logged in users of the application:

<img width="1012" alt="screen shot 2017-01-03 at 12 53 03 am" src="https://cloud.githubusercontent.com/assets/7832803/21597791/bbc9c384-d150-11e6-9d5d-4f27455c8d56.png">

Anonymous users will continue to use the Google Drive submission. However, once #90 (sanitize images (both to handle potentially malicious uploads and to remove metadata)) and #148 (privacy-respecting captcha on this form) are implemented then we can remove the login required decorator from this view and deprecate the Google drive submission form. 

Details:
* Uploads are limited in size and in file type (to images).
* Images are uploaded first to `/tmp` and then to the S3 bucket for OpenOversight. We should eventually have the S3 upload and deletion of the file from `/tmp` be a celery task. The first part (saving to `/tmp`) is quite fast
* A row is inserted in the images table with the date of upload, the hash of the image, the path to the image on S3.
* We disallow the upload of duplicate images in the form by first checking that there is not a matching hash already in the images table in the database. 

After we deploy this, we can go in and upload the non-troll pics from Google drive and the useful pics from our social media scrape to S3 using this form. Troll harder CPD

freddymartinez9 commented:
assigning to myself and @r4v5 for review. I worry about the dangling `tmp/` but perhaps we can chroot it or something else (make read-only) to clean up infosec concerns.

freddymartinez9 commented on OpenOversight/tests/test_utils.py:
```diff
@@ -53,13 +53,7 @@ def test_filter_by_badge_no(mockdata):
-    test_filename = 'test.gif'
-    assert OpenOversight.app.utils.allowed_file(test_filename) == False
+def test_hash_file(mockdata):
+    hash_result = OpenOversight.app.utils.hash_file('bacon')
```
where is the file we are expecting to hash `bacon` actually coming from? the test data directory?

redshiftzero replied:
Ah no, I'll rename this function when I rebase this PR on current develop to be more clear, it should probably be called simply `hash`. Here it's just taking the sha256 of the string `bacon`

redshiftzero replied:
Fixed in commit ab55d39.

freddymartinez9 commented:
Besides my one question and @r4v5's review, I think it should work correctly.

r4v5 commented on OpenOversight/app/main/views.py:
```diff
@@ -74,26 +76,37 @@ def submit_complaint():
+        hash_img = hash_file(image_data)
+        hash_found = Image.query.filter_by(hash_img=hash_img).first()
+
+        if not hash_found:
```
say someone uploads two files named 'Photo01.jpg' with different hashes simultaneously. We run in gunicorn so each thread runs without shared state excepting the filesystem.
`filename = secure_filename('Photo01.jpg')`, then we hash the image(s), then we check the database for hash duplicates, and then we write it out to the filesystem at '/tmp/Photo01.jpg'. Which one gets uploaded?

Also, we're trusting workzeug's `safe_filename`, which is about things that are safe to drop into a folder but not necessarily safe to drop into /tmp itself. What if someone manages to clobber something important-ish (like, say, a socket) in /tmp this way? Can we use the `tempfile` package in stdlib to specify the local filename to use instead?

redshiftzero replied:
Addressed the first comment in commit f7bd0f7 by never using the original filename and always using the hash of the file to store both on the webserver filesystem and on S3.

Addressed this in commit ab636de to use `tempfile` to create a temporary directory and then store the file in there, again never using the original filename from the user (I followed [this guide](https://security.openstack.org/guidelines/dg_using-temporary-files-securely.html) to do so).

r4v5 commented on OpenOversight/app/utils.py:
```diff
@@ -1,11 +1,30 @@
+    s3_client = boto3.client('s3')
+    s3_client.upload_file('/tmp/{}'.format(src_filename),
+                          current_app.config['S3_BUCKET_NAME'],
+                          dest_filename,
```
(loading the s3 web console is a terrible experience with 150k files in the same dir too, trust me)

redshiftzero replied:
Good point - fixed in commit 17536cb.

redshiftzero commented on OpenOversight/app/config.py:
```diff
@@ -34,6 +30,15 @@ class BaseConfig(object):
+    AWS_ACCESS_KEY_ID = os.environ.get('AWS_ACCESS_KEY_ID')
+    AWS_SECRET_ACCESS_KEY = os.environ.get('AWS_SECRET_ACCESS_KEY')
+    AWS_DEFAULT_REGION = 'us-west-2'
+    S3_BUCKET_NAME = 'openoversight'
```
Makes sense, should I then remove these things from here?

r4v5 replied:
No, sorry. I meant that `AWS_DEFAULT_REGION` and `S3_BUCKET_NAME` should be set in .env and therefore not hardcoded to values in `config.py`.

redshiftzero replied:
Ahh, word. Okay changed this in commit 48be6c2

redshiftzero commented on OpenOversight/tests/test_utils.py:
```diff
@@ -53,13 +53,7 @@ def test_filter_by_badge_no(mockdata):
+def test_hash_file(mockdata):
+    hash_result = OpenOversight.app.utils.hash_file('bacon')
+    expected_hash = '9cca0703342e24806a9f64e08c053dca7f2cd90f10529af8ea872afb0a0c77d4'
+    assert hash_result == expected_hash
```
OK so I did almost this but not quite: Added a test for the file storage path in commit 66c2bc9 by monkeypatching the boto3 connection (which actually uncovered a bug which I fixed in 47fdf4b). 

Is the library you mean https://github.com/jubos/fake-s3 which looks like https://github.com/jserver/mock-s3 for python?

redshiftzero commented:
OK guys, I'm finished making changes for now: take a look at these changes and let me know if there are further concerns.

redshiftzero commented:
And thanks for all the great comments!! Made this way better :+1: 💯

r4v5 commented on OpenOversight/app/main/views.py:
```diff
@@ -74,26 +80,49 @@ def submit_complaint():
+            os.umask(SAVED_UMASK)
+
+            # Upload file from local filesystem to S3 bucket and delete locally
+            url = upload_file(safe_local_path, original_filename, new_filename)
```
this line can throw boto exceptions if s3 is down or the credentials are incorrect; do we want to display an informative message to the user when this happens (`flash("Your file could not be uploaded at this time due to a server problem. Please retry again later.")` or the like)? I think if we don't, we'll just show the default nginx unstyled 500.

redshiftzero replied:
Good point, added in fc41ad2

r4v5 commented on requirements.txt:
```diff
@@ -1,5 +1,7 @@
 flask>=0.11.1
 python-dotenv
+mock
```
these can go into `dev-requirements.txt` once #163 lands.

redshiftzero replied:
Fixed in c492804

r4v5 commented on OpenOversight/app/main/views.py:
```diff
@@ -74,26 +80,49 @@ def submit_complaint():
+
+        if not hash_found:
+            # Generate new filename
+            file_extension = original_filename.split('.')[-1]
```
`FileAllowed` in `forms.py` asserts that the file uploaded ends in `{.jpg,.jpeg,.png}` right? Because if not, this could be dicey. Can we throw a test or two in that assert that a file with name ending in .jpg, .jpeg, .png gets accepted and more importantly, one with a name ending in, say, `/../../../../../~/etc/passwd` doesn't, so we capture this requirement in our tests?

redshiftzero replied:
Yep this is true, I've added a couple of tests in 5548ba5 to check that files named "passwd" (due to the way that WTForms validates the file upload form, the file has to exist, so I did this "passwd" check in order to get that test passing both on the dev VM and on Travis) and files with invalid file extensions are not uploaded.

redshiftzero commented:
OK I think I've addressed the remainder of the comments here - I also added some additional unit tests for the login and logout routes. Check it out and let me know 😃

redshiftzero commented:
Yeaaaah!! Thanks y'all 😄
